include/PetriEngine/Structures/StateSet.h

```diff
@@ -24,26 +24,211 @@
 #include "AlignedEncoder.h"
 #include "utils/structures/binarywrapper.h"
 #include "utils/errors.h"
+#include "PetriEngine/PQL/Contexts.h"
 
 
 namespace PetriEngine {
     namespace Structures {
-
         class StateSetInterface
         {
         public:
             StateSetInterface(const PetriNet& net, uint32_t kbound, int nplaces = -1) :
             _nplaces(nplaces == -1 ? net.numberOfPlaces() : nplaces),
-            _encoder(_nplaces, kbound), _net(net)
+            _kbound(kbound), _net(net)
             {
                 _discovered = 0;
-                _kbound = kbound;
                 _maxTokens = 0;
                 _maxPlaceBound = std::vector<uint32_t>(net.numberOfPlaces(), 0);
+            }
+
+            virtual ~StateSetInterface() {}
+
+            const PetriNet& net() { return _net;}
+
+            uint32_t maxTokens() const { return _maxTokens; }
+
+            virtual size_t size() const = 0;
+
+            size_t discovered() const {
+                return _discovered;
+            }
+
+            virtual std::pair<size_t, size_t> getHistory(size_t markingid) = 0;
+
+            const std::vector<MarkVal>& maxPlaceBound() const {
+                return _maxPlaceBound;
+            }
+
+
+        protected:
+            size_t _discovered;
+            size_t _nplaces;
+            uint32_t _kbound;
+            uint32_t _maxTokens;
+            std::vector<uint32_t> _maxPlaceBound;
+            const PetriNet& _net;
+        };
+
+        /**
+         * This class defines the state set for a random walk
+         * It is used by the reachability search to store the states during one walk.
+         */
+        class RandomWalkStateSet : public StateSetInterface
+        {
+        public:
+            RandomWalkStateSet(const PetriNet& net, uint32_t kbound, const PQL::Condition *query, size_t seed, int nplaces = -1) :
+            StateSetInterface(net, kbound, nplaces), _seed(seed)
+            {
+                srand(_seed);
+                _discovered = 1;
+                _initialMarking = new MarkVal[_nplaces];
+                setInitialMarking(net.makeInitialMarking());
+                _nextMarking = new MarkVal[_nplaces];
+                _nextMarking[0] = std::numeric_limits<MarkVal>::max();
+
+                _initializePotencies(_net.numberOfTransitions(), _initPotency);
+                PQL::DistanceContext context(&_net, initialMarking());
+                _initialDistance = query->distance(context);
+            }
+
+            ~RandomWalkStateSet() {
+                // Clean up the markings
+                delete[] _initialMarking;
```

Comment:
Use unique_ptr instead, then the deallocation is automatic

Reply:
Not sure how to do this, I will check